drivers/adc/adc_mchp_g1.c

```diff
@@ -0,0 +1,911 @@
+
+	return ret;
+}
+#endif /* CONFIG_ADC_ASYNC */
+
+static int adc_mchp_init(const struct device *dev)
+{
+	const struct adc_mchp_dev_config *const dev_cfg = dev->config;
+	struct adc_mchp_dev_data *dev_data = dev->data;
+	int8_t ret = 0;
+
+	dev_data->dev = dev;
+	/* Enable the ADC Clock */
+	LOG_DBG("Clock dev: %p, gclk id: %d, mclk id: %d", dev_cfg->adc_clock.clock_dev,
+		(uint32_t)(dev_cfg->adc_clock.gclk_sys), (uint32_t)(dev_cfg->adc_clock.mclk_sys));
+	/* On Global clock for ADC */
+	ret = clock_control_on(dev_cfg->adc_clock.clock_dev, dev_cfg->adc_clock.gclk_sys);
+	if ((ret != 0) && (ret != -EALREADY)) {
+		LOG_ERR("Failed to enable the GCLK for ADC: %d", ret);
+		return ret;
+	}
+
+	/* On Main clock for ADC */
+	ret = clock_control_on(dev_cfg->adc_clock.clock_dev, dev_cfg->adc_clock.mclk_sys);
+	if ((ret != 0) && (ret != -EALREADY)) {
+		LOG_ERR("Failed to enable the MCLK for ADC: %d", ret);
+		return ret;
+	}
+	ret = (ret == -EALREADY) ? 0 : ret;
+
+	/* Get ADC Clock Frequency */
+	ret = clock_control_get_rate(dev_cfg->adc_clock.clock_dev, dev_cfg->adc_clock.gclk_sys,
+				     (uint32_t *)&dev_cfg->freq);
+	if (ret != 0) {
+		LOG_ERR("Failed to get the clock rate for ADC: %d", ret);
+		return ret;
+	}
+
+	adc_set_prescalar(ADC_REGS, dev_cfg->prescaler);
+	pinctrl_apply_state(dev_cfg->pcfg, PINCTRL_STATE_DEFAULT);
+	adc_interrupt_clear(ADC_REGS);
+	dev_cfg->config_func(dev);
+	adc_interrupt_enable(ADC_REGS, 1);
+	adc_controller_enable(ADC_REGS);
+	adc_context_unlock_unconditionally(&dev_data->ctx);
+
+	return ret;
+}
+
+static DEVICE_API(adc, adc_mchp_api) = {
+	.channel_setup = adc_mchp_channel_setup,
+	.read = adc_mchp_read,
+	.ref_internal = ADC_MCHP_VREF_MV,
+#ifdef CONFIG_ADC_ASYNC
+	.read_async = adc_mchp_read_async,
+#endif
+};
+
+/* Macro to Write the calibration values */
+#define ADC_CALIB_WRITE(sw0_word, n, field)                                                        \
+	do {                                                                                       \
+		uint8_t val_##field = (((sw0_word) & FUSES_SW0_WORD_0_ADC##n##_##field##_Msk) >>   \
+				       FUSES_SW0_WORD_0_ADC##n##_##field##_Pos);                   \
+		ADC_REGS->ADC_CALIB |= ADC_CALIB_##field(val_##field);                             \
+	} while (0)
+
+static inline void adc_init_factory_calib_value(const struct device *dev, int n)
+{
+	const volatile fuses_sw0_fuses_registers_t *fuse_reg =
+		(const volatile fuses_sw0_fuses_registers_t *)SW0_ADDR;
+
+	uint32_t sw0_word = (fuse_reg->FUSES_SW0_WORD_0);
+
+	switch (n) {
+	case 0:
+		ADC_CALIB_WRITE(sw0_word, 0, BIASCOMP);
+		ADC_CALIB_WRITE(sw0_word, 0, BIASR2R);
+		ADC_CALIB_WRITE(sw0_word, 0, BIASREFBUF);
+		break;
+	case 1:
+		ADC_CALIB_WRITE(sw0_word, 1, BIASCOMP);
+		ADC_CALIB_WRITE(sw0_word, 1, BIASR2R);
+		ADC_CALIB_WRITE(sw0_word, 1, BIASREFBUF);
+		break;
+	default:
+		/* Invalid ADC instance */
+		LOG_ERR("Invalid ADC index for calibration");
+		break;
+	}
```

Comment:
Should we consider the fail ASAP here ?
BTW, how this **n** will be invalid if it came from devicetree `DT_INST_FOREACH_STATUS_OKAY`?

```c
if (not valid index) {
    LOG_ERR("Invalid ADC index for calibration");
    return;
}

ADC_CALIB_WRITE(sw0_word, n, BIASCOMP);
ADC_CALIB_WRITE(sw0_word, n, BIASR2R);
ADC_CALIB_WRITE(sw0_word, n, BIASREFBUF);
```